lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java

```diff
@@ -53,6 +53,49 @@ public void shouldAddHeaderClaim() throws Exception {
         assertThat(headerJson, JsonMatcher.hasEntry("asd", 123));
     }
 
+    @Test
+    public void shouldReturnBuilderIfNullMapIsProvided() throws Exception {
+        String signed = JWTCreator.init()
+                                  .withHeader(null)
+                                  .sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+    }
+
+    @Test
+    public void shouldOverwriteExistingIfHeadersMapContainsTheSameKey() throws Exception {
+        Map<String, Object> header = new HashMap<String, Object>();
+        header.put("test", 456);
+
+        String signed = JWTCreator.init()
+                                  .withClaim("test", 123)
+                                  .withHeader(header)
+                                  .sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+        String[] parts = signed.split("\\.");
+        String headerJson = new String(Base64.decodeBase64(parts[0]), StandardCharsets.UTF_8);
+        assertThat(headerJson, JsonMatcher.hasEntry("test", 456));
+    }
+
+    @Test
+    public void shouldRemoveHeaderIfTheValueIsNull() throws Exception {
+        Map<String, Object> header = new HashMap<String, Object>();
+        header.put("test", null);
+        header.put("test2", "isSet");
+
+        String signed = JWTCreator.init()
+                                  .withClaim("test", 123)
+                                  .withHeader(header)
+                                  .sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+        String[] parts = signed.split("\\.");
+        String headerJson = new String(Base64.decodeBase64(parts[0]), StandardCharsets.UTF_8);
+        assertThat(headerJson, JsonMatcher.hasEntry("test", null));
```

Comment:
The final JSON object shouldn't have the key at all. Is not correct for it to have the claim present with a null value. e.g. if you convert this back to a Map, the key shouldn't be part of that Map.